File: two_sum_ii_input_array_is_sorted.cpp
/*
167. Two Sum II - Input Array Is Sorted

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int left = 0;
        int right = numbers.size() - 1;
        
        while (left < right) {
            int currentSum = numbers[left] + numbers[right];
            
            if (currentSum == target) {
                return {left + 1, right + 1};
            } else if (currentSum < target) {
                left++;
            } else {
                right--;
            }
        }
        return {};
    }
};


// Test cases
int main() {
    Solution solution;
    
    // Example 1
    vector<int> numbers1 = {2, 7, 11, 15};
    vector<int> result1 = solution.twoSum(numbers1, 9);
    cout << "Example 1: [" << result1[0] << ", " << result1[1] << "]" << endl;  // Expected: [1, 2]
    
    // Example 2
    vector<int> numbers2 = {2, 3, 4};
    vector<int> result2 = solution.twoSum(numbers2, 6);
    cout << "Example 2: [" << result2[0] << ", " << result2[1] << "]" << endl;  // Expected: [1, 3]
    
    // Example 3
    vector<int> numbers3 = {-1, 0};
    vector<int> result3 = solution.twoSum(numbers3, -1);
    cout << "Example 3: [" << result3[0] << ", " << result3[1] << "]" << endl;  // Expected: [1, 2]
    
    // Edge case: two elements
    vector<int> numbers4 = {1, 2};
    vector<int> result4 = solution.twoSum(numbers4, 3);
    cout << "Example 4: [" << result4[0] << ", " << result4[1] << "]" << endl;  // Expected: [1, 2]
    
    // Edge case: negative numbers
    vector<int> numbers5 = {-5, -3, -1, 0, 2, 4};
    vector<int> result5 = solution.twoSum(numbers5, -4);
    cout << "Example 5: [" << result5[0] << ", " << result5[1] << "]" << endl;  // Expected: [1, 3]
    
    // Edge case: duplicates
    vector<int> numbers6 = {1, 2, 3, 3, 4};
    vector<int> result6 = solution.twoSum(numbers6, 6);
    cout << "Example 6: [" << result6[0] << ", " << result6[1] << "]" << endl;  // Expected: [3, 4]
    
    // Edge case: large numbers
    vector<int> numbers7 = {0, 0, 3, 4};
    vector<int> result7 = solution.twoSum(numbers7, 0);
    cout << "Example 7: [" << result7[0] << ", " << result7[1] << "]" << endl;  // Expected: [1, 2]
 
    return 0;
}

File: two_sum_ii_input_array_is_sorted.py
"""
167. Two Sum II - Input Array Is Sorted

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left = 0
        right = len(numbers) - 1
        
        while left < right:
            current_sum = numbers[left] + numbers[right]
            
            if current_sum == target:
                return [left + 1, right + 1]
            elif current_sum < target:
                left += 1
            else:
                right -= 1
        
        return []


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    numbers1 = [2, 7, 11, 15]
    target1 = 9
    print(f"Example 1: {solution.twoSum(numbers1, target1)}")  # Expected: [1, 2]
    
    # Example 2
    numbers2 = [2, 3, 4]
    target2 = 6
    print(f"Example 2: {solution.twoSum(numbers2, target2)}")  # Expected: [1, 3]
    
    # Example 3
    numbers3 = [-1, 0]
    target3 = -1
    print(f"Example 3: {solution.twoSum(numbers3, target3)}")  # Expected: [1, 2]
    
    # Edge case: two elements
    numbers4 = [1, 2]
    target4 = 3
    print(f"Example 4: {solution.twoSum(numbers4, target4)}")  # Expected: [1, 2]
    
    # Edge case: negative numbers
    numbers5 = [-5, -3, -1, 0, 2, 4]
    target5 = -4
    print(f"Example 5: {solution.twoSum(numbers5, target5)}")  # Expected: [1, 3] (-5 + 1 = -4)
    
    # Edge case: duplicates
    numbers6 = [1, 2, 3, 3, 4]
    target6 = 6
    print(f"Example 6: {solution.twoSum(numbers6, target6)}")  # Expected: [3, 4] (3 + 3 = 6)
    
    # Edge case: large numbers
    numbers7 = [0, 0, 3, 4]
    target7 = 0
    print(f"Example 7: {solution.twoSum(numbers7, target7)}")  # Expected: [1, 2] (0 + 0 = 0)
    